Alternative solution nt kernel for long plaintexts

magnumripper opened the pull request:
See #5246 and #5245

This version has only two branches (if configured for length 59) and doesn't lose any optimizations for single-block crypts.

Despite this, it's too slow on AMD w/ old driver.

On nvidia, it's about as fast as the version in #5246.

magnumripper committed:
NT-long-OpenCL: Support lengths up to 125 as a separate format

This way we still got optimal speed for the original nt-opencl format that
support lengths up to 27.

1 block of MD4 is up to 27 characters.
2 blocks is 59, 3 is 91, 4 is 123 and 5 is 125 (due to core max).

As long as we bump it over 27 we don't seem to gain any speed by limiting
it to less than 125 characters.

See #5245

magnumripper committed:
Drop nt-long-opencl and have nt-opencl support any length

magnumripper committed:
Experiment with less diversion

Half the original MD4 block is now placed first, and the end of it comes
last.  This didn't help a lot despite all optimizations are still in place
for single-block crypts - we're only adding two branches.

magnumripper commented:
This version has two branches but no `else` clauses. That makes `unlikely()` less likely to make a difference, right?

solardiz commented:
> it's too slow on AMD w/ old driver.

Are you adding a/b/c/d variables on top of `hash` array (which is normally also in registers)? If so, this increases register pressure, and perhaps you can avoid that.

> This version has two branches but no `else` clauses. That makes `unlikely()` less likely to make a difference, right?

I'd expect it to make a difference anyway. The whole block can be moved out - yes, with an extra branch at the end.

magnumripper commented:
> This version has only two branches (if configured for length 59)

Weird fact: At lengths over 59 we add code for another full MD4 in a `while` loop. But the longer support I use, the faster nvidia gets. It's at best at 125 and significantly slower at 59.

magnumripper commented:
> Are you adding a/b/c/d variables on top of `hash` array (which is normally also in registers)? If so, this increases register pressure, and perhaps you can avoid that.

Well, I need them! For single block, the very last thing that happens is adding `INIT_A..INIT_D` to the values (unless skipping last steps). But for 2nd block on we need to add the original content that was in `hash[0]..hash[3]` so they're stored in `a..d`.

magnumripper committed:
NT-OpenCL: Selective clearing of keybuffer

This is a tad faster than using an initializer as we normally only use 1/5
of the buffer.
